grpc.go

```diff
@@ -56,40 +58,93 @@ type onlineResource struct {
 
 var emptyResp = &gpb.Empty{}
 
-func onlineCPUMem() error {
-	resourceList := []onlineResource{
-		{
-			sysfsOnlinePath: sysfsCPUOnlinePath,
-			regexpPattern:   cpuRegexpPattern,
-		},
-		{
-			sysfsOnlinePath: sysfsMemOnlinePath,
-			regexpPattern:   memRegexpPattern,
-		},
+var onlineCPUMemLock sync.Mutex
+
+const onlineCPUMemWaitTime = 100 * time.Millisecond
+
+const onlineCPUMaxTries = 10
+
+// online resources, nbResources specifies the maximum number of resources to online.
+// If nbResources is -1 then there is no limit, all resources are connected
+// returns the number of resources connected
+func onlineResources(resource onlineResource, nbResources int32) (uint32, error) {
+	files, err := ioutil.ReadDir(resource.sysfsOnlinePath)
+	if err != nil {
+		return 0, err
 	}
 
-	for _, resource := range resourceList {
-		files, err := ioutil.ReadDir(resource.sysfsOnlinePath)
+	var count uint32
+	for _, file := range files {
+		matched, err := regexp.MatchString(resource.regexpPattern, file.Name())
 		if err != nil {
-			return err
+			return count, err
 		}
 
-		for _, file := range files {
-			matched, err := regexp.MatchString(resource.regexpPattern, file.Name())
-			if err != nil {
-				return err
-			}
+		if !matched {
+			continue
+		}
 
-			if !matched {
-				continue
-			}
+		onlinePath := filepath.Join(resource.sysfsOnlinePath, file.Name(), "online")
+		status, err := ioutil.ReadFile(onlinePath)
+		if err != nil {
+			// resource cold plugged
+			continue
+		}
 
-			onlinePath := filepath.Join(resource.sysfsOnlinePath, file.Name(), "online")
+		if strings.Trim(string(status), "\n\t ") == "0" {
 			ioutil.WriteFile(onlinePath, []byte("1"), 0600)
+			count++
+		}
+
+		if nbResources != -1 && count == uint32(nbResources) {
+			return count, nil
 		}
 	}
 
-	return nil
+	return count, nil
+}
+
+func onlineCPUResources(nbCpus uint32) error {
+	resource := onlineResource{
+		sysfsOnlinePath: sysfsCPUOnlinePath,
+		regexpPattern:   cpuRegexpPattern,
+	}
+
+	var count uint32
+	for i := uint32(0); i < onlineCPUMaxTries; i++ {
+		r, err := onlineResources(resource, int32(nbCpus-count))
```

Comment:
Do the directory contents of resource.sysfsOnlinePath change between invocations of this function?  If not do you need to read the directory up to 10 times and check the online files 10 times.  I'm suppose I'm thinking of a situation in which in the first invocation you online 4 out of 5 cpus.  If I read the code correctly you'd then re-read the online files associated with those CPUs up to 9 more times.

Reply:
In systems running low on resources, yes. Read the directory in each iteration is needed